pkg/kubelet/kubelet.go

```diff
@@ -2444,31 +2455,75 @@ func (kl *Kubelet) cleanUpContainersInPod(podID types.UID, exitedContainerID str
 	}
 }
 
-// fastStatusUpdateOnce starts a loop that checks the internal node indexer cache for when a CIDR
-// is applied  and tries to update pod CIDR immediately. After pod CIDR is updated it fires off
-// a runtime update and a node status update. Function returns after one successful node status update.
+// fastStatusUpdateOnce starts a loop that checks if the current state of kubelet + container runtime
+// would be able to turn the node ready, and sync the ready state to the API server as soon as possible.
+// Function returns after the node status update after such event, or when the node is already ready.
 // Function is executed only during Kubelet start which improves latency to ready node by updating
-// pod CIDR, runtime status and node statuses ASAP.
+// kubelet state, runtime status and node statuses ASAP.
 func (kl *Kubelet) fastStatusUpdateOnce() {
+	defer func() {
+		kl.containerRuntimeReadyExpected = true
+	}()
+
 	ctx := context.Background()
-	for {
-		time.Sleep(100 * time.Millisecond)
-		node, err := kl.GetNode()
+
+	nodeIsReady := func(node *v1.Node) bool {
+		for _, c := range node.Status.Conditions {
+			if c.Type == v1.NodeReady {
+				return c.Status == v1.ConditionTrue
+			}
+		}
+
+		klog.ErrorS(nil, "Node does not have NodeReady condition", "node", node)
+		return false
+	}
+
+	var originalNode, node *v1.Node
+
+	start := kl.clock.Now()
+
+	// Sleep() is here so the delay only applies on retry.
+	for ; ; kl.clock.Sleep(100 * time.Millisecond) {
+
+		if kl.clock.Since(start) > nodeReadyGracePeriod {
+			klog.ErrorS(nil, "Node not becoming ready in time after startup")
+			return
+		}
+
+		var err error
+		originalNode, err = kl.GetNode()
 		if err != nil {
 			klog.ErrorS(err, "Error getting node")
 			continue
 		}
-		if len(node.Spec.PodCIDRs) != 0 {
-			podCIDRs := strings.Join(node.Spec.PodCIDRs, ",")
-			if _, err := kl.updatePodCIDR(ctx, podCIDRs); err != nil {
-				klog.ErrorS(err, "Pod CIDR update failed", "CIDR", podCIDRs)
-				continue
-			}
-			kl.updateRuntimeUp()
-			kl.syncNodeStatus()
+
+		if nodeIsReady(originalNode) {
 			return
 		}
+
+		kl.updateRuntimeUp()
```

Comment:
this is the infamous "Container runtime network not ready", "networkReady"" error message, that IIRC it is the runtime merely checking that the cni config file exist. 

This seems to be only CRI calls, so we are still not hitting apiserver